doc/design/state-machine.md

```diff
@@ -0,0 +1,249 @@
+
+Operation: none.
+
+__e<sub>0</sub>__:
+
+Condition: triggered by AG from `Pending` to `Allocated` (e<sub>0</sub> in AG state machine).
+
+Operation: allocate the cell to the AG.
+
+__e<sub>1</sub>__:
+
+Condition: the `Allocated` AG on this is deleted (e<sub>3</sub> in AG state machine).
+
+Operation: release the cell.
+
+__e<sub>2</sub>__:
+
+Condition: triggered by another AG from `Pending` to `Preempting` (i.e., that AG is preempting the `Allocated` AG currently associated with this cell) (e<sub>1</sub> in AG state machine).
+
+Operation: 
+
+The `Allocated` AG on this cell -> `Being preempted` (e<sub>6</sub> in AG state machine);
+
+Release the cell, and then reserve it for the `Preempting` AG.
+
+__e<sub>3</sub>__:
+
+Condition: triggered by another AG from `Pending` to `Preempting` (i.e., that AG cancels the preemption of the `Preempting` AG currently associated with this cell, and continues to preempt the `Being preempted` AG associated with this cell) (e<sub>1</sub> in AG state machine).
+
+Operation:
+
+The original `Preempting` AG on this cell -> `Pending` (e<sub>5</sub> in AG state machine);
+
+Release the cell, and then reserve it for the new `Preempting` AG.
+
+__e<sub>4</sub>__:
+
+Condition: triggered by the `Preempting` AG currently associated with this cell to `Pending` (e<sub>5</sub> in AG state machine) or to `Deleted` (e<sub>7</sub> in AG state machine).
+
+Operation: release the cell, and then allocate it to the `Being preempted` AG on this cell (i.e., the preemption victim).
+
+__e<sub>5</sub>__:
+
+Condition: triggered by AG from `Pending` to `Preempting` (e<sub>1</sub> in AG state machine).
+
+Operation: reserve the cell for the `Preempting` AG.
+
+__e<sub>6</sub>__:
+
+Condition: triggered by another AG from `Pending` to `Preempting` (i.e., that AG cancels the preemption of the `Preempting` AG currently associated with this cell) (e<sub>1</sub> in AG state machine).
+
+Operation:
+
+The original `Preempting` AG on this cell -> `Pending` (e<sub>5</sub> in AG state machine).
+
+Release the cell, and then reserve it for the new `Preempting` AG.
+
+__e<sub>7</sub>__:
+
+Condition: triggered by the `Preempting` AG currently associated with this cell to `Pending` (e<sub>5</sub> in AG state machine) or to `Deleted` (e<sub>7</sub> in AG state machine).
+
+Operation: release the cell.
+
+__e<sub>8</sub>__:
+
+Condition: triggered by (i) there is currently a `Preempting` AG on this cell but another `Allocated` AG is now associated with the cell (e<sub>0</sub> in AG state machine); OR (ii) the `Preempting` AG currently associated with this cell transitions to `Allocated` (e<sub>2</sub> in AG state machine).
+
+Operation: 
+
+For (i): the `Preempting` AG on this cell -> `Pending`  (e<sub>5</sub> in AG state machine); release the cell and then allocate it to the new `Allocated` AG.
+
+For (ii): none.
+
+__e<sub>9</sub>__:
+
+Condition: the pod on this cell of the `Being preempted` AG on the cell is deleted.
```

Comment:
The "the pod on this cell of the `Being preempted` AG on the cell is deleted."  is too hard to read ...

Reply:
But the transition from Reserving to Reserved is pod by pod... Once the pod on a cell is deleted, the cell transitions to Reserved. No need to wait for the whole group to be deleted.
That said, transition all Reserving cells to Reserved when the whole group is deleted may also work. But that would be inconsistent with our implementation